examples/server/server.cpp

```diff
@@ -2266,17 +2266,31 @@ struct server_context {
                 llama_token_data_array cur_p = { slot.ctx_sampling->cur.data(), slot.ctx_sampling->cur.size(), false };
                 result.tok = id;
 
-                const int32_t n_probs = slot.sparams.n_probs;
-                if (slot.sparams.temp <= 0 && n_probs > 0) {
-                    // for llama_sample_token_greedy we need to sort candidates
-                    llama_sample_softmax(ctx, &cur_p);
-                }
+                const size_t n_probs = std::min(cur_p.size, (size_t) slot.sparams.n_probs);
+                if (n_probs > 0) {
+                    const size_t n_considered = slot.ctx_sampling->n_considered;
 
-                for (size_t i = 0; i < std::min(cur_p.size, (size_t) n_probs); ++i) {
-                    result.probs.push_back({
-                        cur_p.data[i].id,
-                        cur_p.data[i].p
-                    });
+                    // Make sure at least n_probs top tokens are at the front of the vector:
+                    if (n_probs > n_considered) {
+                        llama_sample_top_k(ctx, &cur_p, n_probs, 0);
+                    }
+
+                    if (slot.sparams.temp <= 0.0f) {
```

Comment:
Should this check just equality?

```suggestion
                    if (slot.sparams.temp == 0.0f) {
```

For sub-zero temperatures, we calculate the probabilities, though they are never used:

https://github.com/ggerganov/llama.cpp/blob/3af34c1d1b0da47f85b95f60922abeded1cb5d33/common/sampling.cpp#L196-L202

I guess either way makes sense

Reply:
No, I think that it's better to be consistent if at all possible. I changed it to an equality and added a note to the documentation. (I also noticed and fixed a bug where for temperature 0.0f the "top tokens" were wrong.)